tests/activation/test_prompts.py

```diff
@@ -0,0 +1,280 @@
+def platform_check_skip(tmp_path_factory):
+    """Return function triggering skip based on platform & shell."""
+    platform_incompat = "No sane provision for {} on {} yet"
+
+    def check(platform, shell):
+
+        if (sys.platform.startswith("win") and shell in ["bash", "csh", "fish"]) or (
+            sys.platform.startswith("linux") and shell in ["cmd", "powershell"]
+        ):
+            pytest.skip(platform_incompat.format(shell, platform))
+
+        if sys.platform.startswith("win") and shell == "xonsh":
+            pytest.skip("Provisioning xonsh on windows is unreliable")
+
+        if shell == "xonsh" and sys.version_info < (3, 4):
+            pytest.skip("xonsh requires Python 3.4 at least")
+
+        if shell == "powershell":
+            test_ps1 = tmp_path_factory.mktemp("posh_test") / "test.ps1"
+            test_ps1.write_text("echo foo\n")
+
+            if 0 != subprocess.call("powershell -File {}".format(str(test_ps1)), shell=True):
+                pytest.skip("powershell script execution fails; is it enabled?")
+                # Enable with:  PS> Set-ExecutionPolicy -scope currentuser -ExecutionPolicy Bypass -Force;
+                # Disable with: PS> Set-ExecutionPolicy -scope currentuser -ExecutionPolicy Restricted -Force;
+
+    return check
+
+
+@pytest.fixture(scope="module")
+def custom_prompt_root(tmp_path_factory):
+    """Provide Path to root with default and custom venvs created."""
+    root = tmp_path_factory.mktemp("custom_prompt")
+    virtualenv.create_environment(
+        str(root / ENV_CUSTOM), prompt=PREFIX_CUSTOM, no_setuptools=True, no_pip=True, no_wheel=True
+    )
+
+    _, _, _, bin_dir = virtualenv.path_locations(str(root / ENV_DEFAULT))
+
+    bin_dir_name = os.path.split(bin_dir)[-1]
+
+    return root, bin_dir_name
+
+
+@pytest.fixture(scope="module")
+def clean_python_root(clean_python):
+    root = Path(clean_python[0]).resolve().parent
+    bin_dir_name = os.path.split(clean_python[1])[-1]
+
+    return root, bin_dir_name
+
+
+@pytest.fixture(scope="module")
+def get_work_root(clean_python_root, custom_prompt_root):
+    def pick_root(env):
+        if env == ENV_DEFAULT:
+            return clean_python_root
+        elif env == ENV_CUSTOM:
+            return custom_prompt_root
+        else:
+            raise ValueError("Invalid test environment")
+
+    return pick_root
+
+
+@pytest.fixture(scope="module")
+def shell_info():
+    ShellInfo = namedtuple(
+        "ShellInfo",
+        [
+            "execute_cmd",
+            "prompt_cmd",
+            "activate_script",
+            "testscript_extension",
+            "preamble_cmd",
+            "activate_cmd",
+            "deactivate_cmd",
+        ],
+    )
+
+    # execute_cmd, prompt_cmd, activate_script are required.
+    # Defaults here are for testscript_extension, preamble_cmd, activate_cmd, and deactivate_cmd.
+    ShellInfo.__new__.__defaults__ = ("", "", "source ", "deactivate")
```

Comment:
is this the official way? 😄 haven't seen it yet 👍

Reply:
Based on an SO post, I'm not sure there *is* an official way for <3.7.... 3.7 has a `defaults=` argument to the constructor, which AFAIK is canonical.

I can re-implement without this if desired, it'll just be more verbose with all of the replicated default values.

Reply:
Converted to fully explicit instantiation of all `ShellInfo` instances in 5e5b5c1.